pkg/cloudprovider/providers/gce/gce_address_manager.go

```diff
@@ -0,0 +1,155 @@
+	// Get the address in case it was orphaned earlier
+	addr, err := am.svc.GetBetaRegionAddress(am.lbName, am.region)
+	if err != nil && !isNotFound(err) {
+		return err
+	}
+
+	if addr != nil {
+		// If address exists, check if the address IP is different from what's expected.
+		if addr.Address != am.targetIP || addr.AddressType != string(am.addressType) {
+			glog.V(3).Infof("%v: existing address %q has IP %v Type %v which does not match targeted IP %v Type %v. Attemping to delete.", am.logPrefix, addr.Name, addr.Address, addr.AddressType, am.targetIP, am.addressType)
+			err := am.svc.DeleteRegionAddress(addr.Name, am.region)
+			if err != nil {
+				if isNotFound(err) {
+					glog.V(3).Infof("%v: address %q was not found. Ignoring.", am.logPrefix, addr.Name)
+				} else {
+					return err
+				}
+			} else {
+				glog.V(3).Infof("%v: successfully deleted previous address %q", am.logPrefix, addr.Name)
+			}
+		} else {
+			glog.V(3).Infof("%v: address %q already reserves targeted IP %v of type %v. No further action required.", am.logPrefix, addr.Name, am.targetIP, am.addressType)
+			return nil
+		}
+	}
+
+	return am.ensureAddressReservation()
+}
+
+// ReleaseAddress will release the address if it's owned by the controller.
+func (am *addressManager) ReleaseAddress() error {
+	if !am.tryRelease {
+		glog.V(3).Infof("%v: not attempting release of address %v.", am.logPrefix, am.targetIP)
+		return nil
+	}
+
+	glog.V(3).Infof("%v: releasing address %v named %q", am.logPrefix, am.targetIP, am.lbName)
+	// Controller only ever tries to unreserve the address named with the load balancer's name.
+	err := am.svc.DeleteRegionAddress(am.lbName, am.region)
+	if err != nil {
+		if isNotFound(err) {
+			glog.Warningf("%v: address %q was not found. Ignoring.", am.logPrefix, am.targetIP, am.lbName)
+			return nil
+		}
+
+		return err
+	}
+
+	glog.V(3).Infof("%v: successfully released IP %v named %q", am.logPrefix, am.targetIP, am.lbName)
+	return nil
+}
+
+func (am *addressManager) ensureAddressReservation() error {
+	// Try reserving the IP with controller-owned address name
+	newAddr := &computebeta.Address{
+		Name:        am.lbName,
+		Description: fmt.Sprintf(`{"kubernetes.io/service-name":"%s"}`, am.serviceName),
+		Address:     am.targetIP,
+		AddressType: string(am.addressType),
+		Subnetwork:  am.subnetURL,
+	}
+
+	err := am.svc.ReserveBetaRegionAddress(newAddr, am.region)
+	if err != nil {
+		if !isHTTPErrorCode(err, http.StatusConflict) {
+			return err
+		}
+	} else {
+		glog.V(3).Infof("%v: successfully reserved IP %v with name %q", am.logPrefix, am.targetIP, newAddr.Name)
+		return nil
+	}
+
+	glog.V(3).Infof("%v: could not reserve IP %v due to err: %v", am.logPrefix, am.targetIP, err)
+
+	// Reserving the address failed due to a conflict. The address manager just checked that no address
+	// exists with the lbName, so it may belong to the user.
+	addr, err := am.svc.GetBetaRegionAddressByIP(am.region, am.targetIP)
```

Comment:
I assume the controller will need to verify that the "user-owned" address has all other desired attributes (e.g., type, tier, etc) out of band. And I'll have to verify the existing ephemeral forwarding-rule IP too. Maybe we should just add a helper function for doing that later...

Reply:
Yes, let's discuss this in person. I'm not sure what action we want to take when the IP matches but the attributes don't. In my opinion, we raise an error and stop syncing.